smb/smb.go

```diff
@@ -754,6 +816,554 @@ type QueryDirectoryRes struct {
+	OwnerSid    *SID
+	GroupSid    *SID
+	Sacl        *PACL
+	Dacl        *PACL
+}
+
+type PACL struct {
+	AclRevision uint16
+	AclSize     uint16
+	AceCount    uint32
+	ACLS        []ACE
+}
+
+// MS-DTYP Section 2.4.4.1 ACE_HEADER
+type ACEHeader struct {
+	Type  byte
+	Flags byte
+	Size  uint16 //Includes header size?
+}
+
+type ACE struct {
+	Header ACEHeader
+	Mask   uint32
+	Sid    SID //Must be multiple of 4
+}
+
+type SID struct {
+	Revision       byte
+	NumAuth        byte
+	Authority      []byte
+	SubAuthorities []uint32
+}
+
+const (
+	securityDescriptorOwner = "owner"
+	securityDescriptorGroup = "group"
+	securityDescriptorDACL  = "dacl"
+	securityDescriptorSACL  = "sacl"
+	securityDescriptorEnd   = "end"
+)
+
+func (self *SecurityDescriptor) MarshalBinary() (ret []byte, err error) {
+	w := bytes.NewBuffer(ret)
+	ptrBuf := make([]byte, 0)
+	// Order: 1. SACL, 2. DACL, 3. Owner, 4. Group
+	bufOffset := uint32(20)
+
+	if self.Sacl != nil {
+		sBuf, err := self.Sacl.MarshalBinary()
+		if err != nil {
+			log.Errorln(err)
+			return nil, err
+		}
+		ptrBuf = append(ptrBuf, sBuf...)
+		self.Control |= SecurityDescriptorFlagSP
+		self.OffsetSacl = bufOffset
+		bufOffset += uint32(len(sBuf))
+	}
+	if self.Dacl != nil {
+		dBuf, err := self.Dacl.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		ptrBuf = append(ptrBuf, dBuf...)
+		self.Control |= SecurityDescriptorFlagDP
+		self.OffsetDacl = bufOffset
+		bufOffset += uint32(len(dBuf))
+	}
+
+	if self.OwnerSid != nil {
+		oBuf, err := self.OwnerSid.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		ptrBuf = append(ptrBuf, oBuf...)
+		self.OffsetOwner = bufOffset
+		bufOffset += uint32(len(oBuf))
+	}
+
+	if self.OffsetGroup != 0 {
+		gBuf, err := self.GroupSid.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		ptrBuf = append(ptrBuf, gBuf...)
+		self.OffsetGroup = bufOffset
+	}
+
+	// Encode revision
+	err = binary.Write(w, binary.LittleEndian, self.Revision)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+	// Encode control
+	err = binary.Write(w, binary.LittleEndian, self.Control)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+	// Encode  OffsetOwner
+	err = binary.Write(w, binary.LittleEndian, self.OffsetOwner)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+	// Encode  OffsetGroup
+	err = binary.Write(w, binary.LittleEndian, self.OffsetGroup)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+	// Encode  OffsetSacl
+	err = binary.Write(w, binary.LittleEndian, self.OffsetSacl)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+	// Encode  OffsetDacl
+	err = binary.Write(w, binary.LittleEndian, self.OffsetDacl)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+
+	// Encode serialized Owner, Group, Sacl and Dacl
+	_, err = w.Write(ptrBuf)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+
+	return w.Bytes(), nil
+}
+
+func (self *SecurityDescriptor) UnmarshalBinary(buf []byte) (err error) {
```

Comment:
Implement BinaryMarshallable interface
`func (self *SecurityDescriptor) UnmarshalBinary(buf []byte) (err error) {`
should be
`func (self *SecurityDescriptor) UnmarshalBinary(buf []byte, meta *encoder.Metadata) (err error) {`